src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs

```diff
@@ -332,127 +372,134 @@ private void WatchForFileSystemEventsThreadStart(object arg)
                     return;
                 }
 
-                // Since renames come in pairs, when we find the first we need to search for the next one. Once we find it, we'll add it to this
-                // list so when the for-loop comes across it, we'll skip it since it's already been processed as part of the original of the pair.
-                List<FSEventStreamEventId> handledRenameEvents = null;
-                Memory<char>[] events = new Memory<char>[numEvents.ToInt32()];
-                ProcessEvents();
-
-                for (long i = 0; i < numEvents.ToInt32(); i++)
+                ExecutionContext.Run(context, delegate (object state)
```

Comment:
Do you know which ExecutionContext implementation Mono now uses?  Presumably it's using the core implementation, in which case this is good  If, however, it's using the netfx-based implementation, this will likely cause problems, as the netfx implementation only allows an individual ExecutionContext instance to be used to run once, so this would need a CreateCopy.  I took a quick peek at Mono, and it seems like it's using the core implementation, so this should be fine.

Reply:
Actually Mono uses reference-source implementation of ExecutionContext. I faced the issue you mentioned during testing in VS for Mac and trying to execute more than one operation on the tracked file. ExecutionContext.Run sets isNewCapture to false and next time throws an exception. Trying to avoid it I just recapture execution context. If it's more correct to use CreateCopy method I'll do so.